### DISCUSSION TITLE
How to fuzzy find git hunk content?

### DISCUSSION THREAD

**Discussion by MrConnorKenway** [ibhagwan/fzf-lua#1598] [Q&A] [answered]
I want to fuzzy search all hunks (or unstaged hunks if an optional argument is provided) inside whole project, and preview them using something like git diff to show the diff of every git hunk entry. Also, when there is no input, I can simply navigate all hunks in the previewer. Fzflua's git status does not fully match my purpose because I can only fuzzy search the file name rather than the hunk content. I've tried https://github.com/radyz/telescope-gitsigns, but it only previews gitsigns in current buffer and doesn't search the content. Is it possible to implement it using a few lines of lua code inside my config?

**Comment by ibhagwan** [comment]
You can easily fuzzy the current diff using `fzf_exec('git diff', { <opts> })`, as for the preview you’ll need to find a command that can take a part of the diff line and preview it using a command + fzf field index expressions, I.e. something like `git show {1}`.

If you can find/craft git command that outputs something like this `file:line:col: <diff line>` you can reuse the grep pickers with `{ cmd = … }` to fuzzy over the diff seamlessly with a preview.

**Reply by ibhagwan** [reply]
Take a look at this discussion, the user creates a simple `diff2vimgrep` script, perhaps you can reuse that and then feed this command to grep as `grep.({ raw_cmd = 'git diff | diff2vimgrep' })` (using raw_cmd as we aren’t sending an underlying regex):
https://unix.stackexchange.com/questions/628695/grep-within-a-git-diff-but-show-file-and-line-number

**Reply by MrConnorKenway** [reply]
I've tried `diff2vimgrep`, but the line number messed up if the hunk involves modification, and it cannot search deleted content. So I use another script which outputs something like:
```
nvim/lazy-lock.json:35:1: +  "nvim-treesitter": { "branch": "master", "commit": "54a29bf99d5ccb2de512906279686457d699f8e5" },
nvim/lazy-lock.json:36:1: +  "nvim-web-devicons": { "branch": "master", "commit": "1736cb81bdc42e1ee206081a161bdb2af68fb7a1" },
nvim/lazy-lock.json:50:1: -  "vim-fugitive": { "branch": "master", "commit": "320b18fba2a4f2fe3c8225c778c687e0d2620384" },
nvim/lazy-lock.json:50:1: -  "vim-illuminate": { "branch": "master", "commit": "5eeb7951fc630682c322e88a9bbdae5c224ff0aa" }
nvim/lazy-lock.json:50:1: +  "vim-fugitive": { "branch": "master", "commit": "fcb4db52e7f65b95705aa58f0f2df1312c1f2df2" },
nvim/lazy-lock.json:51:1: +  "vim-illuminate": { "branch": "master", "commit": "895351de7bbf93c1802f9087a53ee7a056f6b375" }
```
But I cannot get the bat previewer move to corresponding line. What am I missing?
```lua
      {
        'gh',
        function()
          local fzf_lua = require('fzf-lua')
          local opts = {}

          opts.file_icons = true
          opts.color_icons = true
          -- setup default actions for edit, quickfix, etc
          opts.actions = fzf_lua.defaults.actions.files
          opts.previewer = 'bat_native'
          opts.fn_transform = function(x)
            return fzf_lua.make_entry.file(x, opts)
          end
          fzf_lua.fzf_exec('bash diffn.sh', opts)
        end
      },
```
If I use `fzf_lua.grep({raw_cmd='bash diffn.sh'})`, the previewer can move along with `C-n` `C-p`, but I cannot perform regex search.

**Reply by MrConnorKenway** [reply]
After adding `require('fzf-lua.core').set_fzf_field_index(opts)` before `fzf_exec`, the previewer is functional now. Not sure if this is the right answer though.

Sorry but I want to ask another question, in the above lua code, how can I change to regex search inside `fzf_exec` using `ctrl-g`?

**Comment by MrConnorKenway** [comment, marked as answer]
@ibhagwan Thank you for your guidance <3, I managed to implement this feature. However, when I press `ctrl-g`, the whole screen will become flickering. By the way, would you be interested in making this a built-in feature of fzf-lua?

```lua
      {
        'gh',
        function()
          local M = {}

          M.toggle_regex = function(_, opts)
            opts.__ACT_TO {
              resume = true,
            }
          end

          local fzf_lua = require('fzf-lua')
          local common_opts = {
            winopts = { title = ' Git Hunks ', title_pos = 'center' },
            actions = fzf_lua.defaults.actions.files,
            file_icons = true,
            color_icons = true,
            previewer = 'bat_native',
            field_index_expr = '{1}',
            line_field_index = '{2}',
            multiprocess = true,
            fzf_opts = {
              ["--multi"] = true
            }
          }
          common_opts.actions['ctrl-g'] = M.toggle_regex

          local fn_transform = function(x)
            return fzf_lua.make_entry.file(x, common_opts)
          end

          -- diffn parses git diff and produces line number for each hunk
          local diffn = function(fzf_cb, live_query)
            local diff_text = vim.fn.system('git diff')
            local file_name = nil
            local line_number = nil
            local counter = 0

            for line in diff_text:gmatch('([^\n]+)') do
              if counter < 3 then
                counter = counter + 1
                goto continue
              end

              if counter == 3 then
                -- +++ b/
                file_name = line:match('^%+%+%+ b/(.*)')
                counter = counter + 1
                goto continue
              end

              local char = line:sub(1, 1)

              if char == '-' then
                if not live_query or line:match(live_query) then
                  fzf_cb(fn_transform(file_name) .. ':' .. line_number .. ':' .. ' \27[31m' .. line .. '\27[m')
                end
                goto continue
              end

              if char == '+' then
                if not live_query or line:match(live_query) then
                  fzf_cb(fn_transform(file_name) .. ':' .. line_number .. ':' .. ' \27[32m' .. line .. '\27[m')
                end
                line_number = line_number + 1
                goto continue
              end

              if char == ' ' then
                line_number = line_number + 1
                goto continue
              end

              local new_line_number = line:match('^@@ %-%d+,%d+ %+(%d+),%d+ @@')
              if new_line_number then
                line_number = new_line_number
                goto continue
              end

              counter = 1
              ::continue::
            end

            fzf_cb()
          end

          M.fuzzy_hunks = function(opts)
            opts = opts or {}
            opts = vim.tbl_deep_extend('keep', opts, common_opts)
            opts.__ACT_TO = M.lgrep_hunks
            opts.prompt = 'fuzzy> '
            opts = require('fzf-lua.config').normalize_opts(opts, {}, "hunk")

            fzf_lua.fzf_exec(function(fzf_cb)
              diffn(fzf_cb)
            end, opts)
          end

          M.lgrep_hunks = function(opts)
            opts = opts or {}
            opts = vim.tbl_deep_extend('keep', opts, common_opts)
            opts.__ACT_TO = M.fuzzy_hunks
            opts.prompt = 'lgrep> '
            opts.exec_empty_query = true
            opts = require('fzf-lua.config').normalize_opts(opts, {}, "hunk")

            fzf_lua.fzf_live(
              function(query)
                return function(fzf_cb)
                  diffn(fzf_cb, query)
                end
              end,
              opts
            )
          end

          M.fuzzy_hunks()
        end
      }
```

**Reply by ibhagwan** [reply]
>  I managed to implement this feature. 

Nice work :)

> However, when I press ctrl-g, the whole screen will become flickering

That’s because fzf-lua is closing the window and recreating it, to prevent that just wrap your function in a table, that should tell fzf-lua to reuse the window.:

```lua
common_opts.actions['ctrl-g'] = { M.toggle_regex }
```

> By the way, would you be interested in making this a built-in feature of fzf-lua?

Not ATM as this would require a lot of testing, performance testing and all kinds of edge cases with the native previewer probably, I don’t have the capacity for that right now :)

**Reply by MrConnorKenway** [reply]
Damn it, I used to think that the curly braces are redundant and removed them :(
Thank you for your precious time! Merry Christmas in advance!

**Reply by ibhagwan** [reply]
> Damn it, I used to think that the curly braces are redundant and removed them :( Thank you for your precious time! Merry Christmas in advance!

That’s because it’s a mere “convenience” format, the more explicit would be `{ fn = …, noclose = true }`

Happy holidays 🎉

**Reply by ibhagwan** [reply]
FYI @MrConnorKenway, similar discussion slightly different solution in #1640, you might be interested:

> diff2vimgrep
```sh
#!/usr/bin/env bash

set -eEuCo pipefail

# source: https://stackoverflow.com/a/33249416/51209
cat | awk '
  match($0, "^@@ -([0-9]+),([0-9]+) [+]([0-9]+),([0-9]+) @@", a) {
    right=a[3]
  }
  match($0, "^+++ b/([^ ]+)", b) {
    path=b[1]
  }
  { right++ }
  /^\+\+\+ / { next }
  ! /^\+/ { next }

  { line=substr($0, 2) }
  # match(line, "^$") { next }

  { printf "%s:%s:1:%s\n", path, right-2, line }
'
```

fzf-lua command is simpler as a result:
> Remove `debug=true` once you're done experimenting
```lua
:lua require("fzf-lua").grep({raw_cmd=[[{ git diff main... & git diff; } | diff2vimgrep | sort -u]],debug=true})
```

![image](https://github.com/user-attachments/assets/47e37932-3c3e-4968-92a6-57dac3db128d)

**Reply by MrConnorKenway** [reply]
> FYI @MrConnorKenway, similar discussion slightly different solution in #1640, you might be interested:

Thank you for your update. I have thought the similar thing, but I prefer not to introduce additional dependencies like installing a script into my PATH, so I ended up doing stuffs in lua. Anyway, appreciate your kindness.

**Comment by ibhagwan** [comment]
FYI @MrConnorKenway, in case you're still using fzf-lua I added a modified version of your code as a builtin picker in fzf-lua :-)
```lua
:FzfLua  git_hunks
:FzfLua  git_hunks ref=HEAD^
:FzfLua  git_hunks ref=dev_branch
```

https://github.com/ibhagwan/fzf-lua/commit/01083f73af7128cfea637de99675e4cc03601d48

**Reply by MrConnorKenway** [reply]
Cool! I have roughly reviewed this commit, and it seems that the parser ignores patterns described in https://git-scm.com/docs/diff-format#_combined_diff_format. For example:
```diff
@@@ <from-file-range> <from-file-range> <to-file-range> @@@
```

**Reply by ibhagwan** [reply]
> Cool! I have roughly reviewed this commit, and it seems that the parser ignores patterns described in https://git-scm.com/docs/diff-format#_combined_diff_format. For example:
> 
> ```diff
> @@@ <from-file-range> <from-file-range> <to-file-range> @@@
> ```

I copied the line march from your code above, you’re welcome to submit a fix PR :)

Otherwise I’ll handle this tomorrow.
